packets2/connack.go

```diff
@@ -0,0 +1,84 @@
+package packets2
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	pkts "github.com/energomonitor/bisquitt/packets"
+)
+
+const connackHeaderLength uint16 = 6
+const connackSessionPresentBit uint8 = 1
+
+type Connack struct {
+	pkts.Header
+	PacketV2
+	// Flags
+	SessionPresent bool
+	// Fields
+	SessionExpiryInterval    uint32
+	ReasonCode               ReasonCode
+	AssignedClientIdentifier string
+}
+
+func NewConnack(sessExpiry uint32, reasonCode ReasonCode, assignedClientID string,
+	sessPresent bool) *Connack {
+	p := &Connack{
+		Header:                   *pkts.NewHeader(pkts.CONNACK, 0),
+		SessionPresent:           sessPresent,
+		SessionExpiryInterval:    sessExpiry,
+		ReasonCode:               reasonCode,
+		AssignedClientIdentifier: assignedClientID,
+	}
+	p.computeLength()
+	return p
+}
+
+func (p *Connack) computeLength() {
+	assignedClientIDLength := uint16(len(p.AssignedClientIdentifier))
+	p.Header.SetVarPartLength(connackHeaderLength + assignedClientIDLength)
+}
+
+func (p *Connack) decodeFlags(b byte) {
+	p.SessionPresent = (b & connackSessionPresentBit) == connackSessionPresentBit
+}
+
+func (p *Connack) encodeFlags() byte {
+	var b byte
+	if p.SessionPresent {
+		b |= connackSessionPresentBit
+	}
+	return b
+}
+
+func (p *Connack) Pack() ([]byte, error) {
+	buf := p.Header.PackToBuffer()
+
+	_ = buf.WriteByte(byte(p.ReasonCode))
+	_ = buf.WriteByte(p.encodeFlags())
+	_, _ = buf.Write(pkts.EncodeUint32(p.SessionExpiryInterval))
```

Comment:
In the spec, the Session Expiry Interval field is marked as optional. The situation when it is missing is equal to a situation where it has a value 0. I think it would be better if it is not emitted in that case.

There is a precedent [with the `DISCONNECT` packet](https://github.com/energomonitor/bisquitt/blob/v2-connection/packets1/disconnect.go#L44-L46) in `packets1`.

Reply:
The problem is that the optional `Session Expiry Interval` is followed by also optional `Assigned Client Identifier` and the specification does not mention any mechanism how to signalize which of the optional fields is present. I think this is a bug in the specification and therefore I (temporarily?) consider `Session Expiry Interval` mandatory.

This should probably be explicitly stated somewhere in the code.

Reply:
P.S. A similar but different situation is with the `Pingreq` packet where there are two optional fields `Max Messages` and `Client Identifier`.

The difference is that I suppose **both** fields are either present or both are missing. Because the `Pingreq` packet is either used as a "Are you alive?" or "I'm going to awake state".

The specification kind of support this understanding:

[Max Messages] is optional, only included when a sleeping client goes to the awake state. [...] [Client Identifier] is optional and is included by a “sleeping” client when it goes to the “awake” state [2.1.22.2 and 2.1.22.3 in WD20]

Therefore, the implementation is different in `Pingreq`:
https://github.com/energomonitor/bisquitt/blob/v2-connection/packets2/pingreq.go#L37

Reply:
> The problem is that the optional `Session Expiry Interval` is followed by also optional `Assigned Client Identifier` and the specification does not mention any mechanism how to signalize which of the optional fields is present.

I see.

> This should probably be explicitly stated somewhere in the code.

Definitely, please add a comment.